graph/src/data/store/scalar.rs

```diff
@@ -193,30 +192,58 @@ impl bigdecimal::ToPrimitive for BigDecimal {
     }
 }
 
+impl stable_hash_legacy::StableHash for BigDecimal {
+    fn stable_hash<H: stable_hash_legacy::StableHasher>(
+        &self,
+        mut sequence_number: H::Seq,
+        state: &mut H,
+    ) {
+        let (int, exp) = self.as_bigint_and_exponent();
+        // This only allows for backward compatible changes between
+        // BigDecimal and unsigned ints
+        stable_hash_legacy::StableHash::stable_hash(&exp, sequence_number.next_child(), state);
+        stable_hash_legacy::StableHash::stable_hash(&BigInt(int), sequence_number, state);
+    }
+}
+
 impl StableHash for BigDecimal {
-    fn stable_hash<H: StableHasher>(&self, mut sequence_number: H::Seq, state: &mut H) {
+    fn stable_hash<H: stable_hash::StableHasher>(&self, field_address: H::Addr, state: &mut H) {
         let (int, exp) = self.as_bigint_and_exponent();
         // This only allows for backward compatible changes between
         // BigDecimal and unsigned ints
-        exp.stable_hash(sequence_number.next_child(), state);
-        BigInt(int).stable_hash(sequence_number, state);
+        StableHash::stable_hash(&exp, field_address.child(0), state);
```

Comment:
The comment says this allows for backward compatible changes, but this implementation does not provide backward compatibility for schema changes as advertised.

One of the defining features of the `stable-hash` crate is that it allows (limited) schema evolution without breaking existing hashes. Adding new fields to a structure is one allowable change, but so is changing between integer types. For example, you can change from a `u8` to an `i16` since the new type contains the old type. Changing from `u32` -> `BigInt` -> `BigDecimal` should also be supported.

This implementation does not allow for those schema evolutions, because it uses `field_adddress.child(1)` for the data in the `BigInt`.

Using the new `stable-hash` API, we can not only support a conversion to `BigDecimal` from an unsigned int (as the comment suggests), but we can support conversion from signed ints as well!

We happen to know that ints are written using `field_address` for a trimmed array of the little endian bytes, and `.child(0)` for the negative sign (if any). So using `.child(1)` for the exponent will be backward compatible without causing collisions.

Typically this kind of assumption would be dangerous, but it's reasonable in this case because `stable-hash` _can't_ change this implementation detail without breaking the contract.

Reply:
I made some changes here but not sure this is what you suggested or if you wanted something like AsInt to be used. 

If you could let me know exactly what the code should look like it might be easier

Reply:
```rust
// This implementation allows for backward compatible changes from integers (signed or unsigned)
// when the exponent is zero.
let (int, exp) = self.as_bigint_and_exponent();
exp.stable_hash(field_address.child(1));
// Normally it would be a red flag to pass field_address in after having used a child slot.
// But, we know the implementation of StableHash for BigInt will not use child(1) and that
// it will not in the future due to having no forward schema evolutions for ints and the
// stability guarantee.
//
// For reference, ints use child(0) for the sign and write the little endian bytes to the parent slot.
int.stable_hash(field_address);
```

Reply:
Can we write a unit test to verify backward compatibility to make sure?